packages/dd-trace/src/debugger/devtools_client/snapshot/collector.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const session = require('../session')
+
+const LEAF_SUBTYPES = new Set(['date', 'regexp'])
+const ITERABLE_SUBTYPES = new Set(['map', 'set', 'weakmap', 'weakset'])
+
+module.exports = {
+  getRuntimeObject: getObject
+}
+
+// TODO: Can we speed up thread pause time by calling mutiple Runtime.getProperties in parallel when possible?
+// The most simple solution would be to swich from an async/await approach to a callback based approach, in which case
+// each lookup will just finish in its own time and traverse the child nodes when the event loop allows it.
+// Alternatively, use `Promise.all` or something like that, but the code would probably be more complex.
+
+async function getObject (objectId, maxDepth, depth = 0) {
```

Comment:
Correct. The snapshot collection is divided into two parts:

1. The first part happens while the target thread is paused and so we want to limit what's going on here to the bare minimum. `maxDepth` is useful in this case, because we can limit the time spend walking the object structure.
2. The second part happens after the target thread has resumed. This is where we, among other things, truncate all strings that was collected in the first part.

Reply:
I see. that is very cool logic. but I think there is missing logic on step #1 and optimization we can have.

Because the capture does `traverseGetPropertiesResult` -> we should also pass maxFields, if you have an object that have more then maxFields -> we don't need to traverse all those fields only the first maxFields and add notCaptureReason that we skipped few fields.

I assume that would be part of the future PRs when add support for those.

Reply:
> I assume that would be part of the future PRs when add support for those.

Correct.